openmls/src/extensions/mod.rs

```diff
@@ -166,6 +166,200 @@ pub enum Extension {
+    }
+
+    // ---------------------------------------------------------------------------------------------
+
+    /// Add an extension to the extension list.
+    ///
+    /// Returns an error when there already is an extension with the same extension type.
+    pub fn add(&mut self, extension: Extension) -> Result<(), InvalidExtensionError> {
+        if !self.contains(extension.extension_type()) {
+            self.inner.push(extension);
+            Ok(())
+        } else {
+            Err(InvalidExtensionError::Duplicate)
+        }
+    }
+
+    /// Add an extension to the extension list (or replace an existing one.)
+    ///
+    /// Returns the replaced extension (if any).
+    pub fn add_or_replace(&mut self, extension: Extension) -> Option<Extension> {
+        let replaced = self.remove(extension.extension_type()).ok();
+        self.inner.push(extension);
+        replaced
+    }
+
+    /// Remove an extension from the extension list.
+    ///
+    /// Returns the removed extension or an error when there is no extension with the given extension type.
+    pub fn remove(
+        &mut self,
+        extension_type: ExtensionType,
+    ) -> Result<Extension, InvalidExtensionError> {
+        match self
+            .inner
+            .iter()
+            .position(|e| e.extension_type() == extension_type)
+        {
+            Some(position) => Ok(self.inner.remove(position)),
+            None => Err(InvalidExtensionError::NotFound),
+        }
+    }
+
+    /// Replace an extension in the extension list.
+    ///
+    /// Returns the replaced extension or an error when there is no extension with the given extension type.
+    #[cfg(any(feature = "test-utils", test))]
+    pub fn replace(&mut self, extension: Extension) -> Result<Extension, InvalidExtensionError> {
+        let mut extension = extension;
+
+        match self
+            .inner
+            .iter()
+            .position(|e| e.extension_type() == extension.extension_type())
+        {
+            Some(position) => {
+                std::mem::swap(&mut self.inner[position], &mut extension);
+                Ok(extension)
+            }
+            None => Err(InvalidExtensionError::NotFound),
+        }
+    }
+
+    // ---------------------------------------------------------------------------------------------
+
+    /// Return true if (and only if) the extension list contains the extension with the given type.
+    pub fn contains(&self, extension_type: ExtensionType) -> bool {
+        self.inner
+            .iter()
+            .any(|e| e.extension_type() == extension_type)
+    }
+
+    /// Check that the candidate extension list is valid.
+    ///
+    /// Specifically, the candidate list must ...
+    ///
+    /// * ... not contain duplicate extension types (ValSem012) ...
+    ///
+    /// ... to be valid.
+    pub fn validate(candidate: &[Extension]) -> Result<(), InvalidExtensionError> {
+        let mut map = BTreeSet::new();
```

Comment:
The operations performed on the internal vector would all be more efficient when you'd always use a set. You also could get rid of this function. Is there a reason not to use a set for `inner`?

Reply:
As discussed privately, using a set would require a wrapper around `Extension` that implements a custom `Ord` and we could use a `HashMap` instead.